src/tag.rs

```diff
@@ -58,183 +81,231 @@ impl Tag {
-    pub const GRAPHIC_STRING: Self = Tag(25);
+    pub const GRAPHIC_STRING: Self = Tag([25, 0, 0, 0]);
 
     /// The tag for the VisibleString type, UNIVERSAL 26.
-    pub const VISIBLE_STRING: Self = Tag(26);
+    pub const VISIBLE_STRING: Self = Tag([26, 0, 0, 0]);
 
     /// The tag for the GeneralString type, UNIVERSAL 27.
-    pub const GENERAL_STRING: Self = Tag(27);
+    pub const GENERAL_STRING: Self = Tag([27, 0, 0, 0]);
 
     /// The tag for the UniversalString type, UNIVERSAL 28.
-    pub const UNIVERSAL_STRING: Self = Tag(28);
+    pub const UNIVERSAL_STRING: Self = Tag([28, 0, 0, 0]);
 
     /// The tag for the BMPString type, UNIVERSAL 29.
-    pub const BMP_STRING: Self = Tag(29);
+    pub const BMP_STRING: Self = Tag([29, 0, 0, 0]);
 
     /// The tag context specific tag [0].
-    pub const CTX_0: Self = Tag(Tag::CONTEXT_SPECIFIC);
+    pub const CTX_0: Self = Tag([Tag::CONTEXT_SPECIFIC, 0, 0, 0]);
 
     /// The tag context specific tag [1].
-    pub const CTX_1: Self = Tag(Tag::CONTEXT_SPECIFIC | 1);
+    pub const CTX_1: Self = Tag([Tag::CONTEXT_SPECIFIC | 0x01, 0, 0, 0]);
 
     /// The tag context specific tag [2].
-    pub const CTX_2: Self = Tag(Tag::CONTEXT_SPECIFIC | 2);
+    pub const CTX_2: Self = Tag([Tag::CONTEXT_SPECIFIC | 0x02, 0, 0, 0]);
 
     /// The tag context specific tag [3].
-    pub const CTX_3: Self = Tag(Tag::CONTEXT_SPECIFIC | 3);
+    pub const CTX_3: Self = Tag([Tag::CONTEXT_SPECIFIC | 0x03, 0, 0, 0]);
 
     /// The tag context specific tag [4].
-    pub const CTX_4: Self = Tag(Tag::CONTEXT_SPECIFIC | 4);
+    pub const CTX_4: Self = Tag([Tag::CONTEXT_SPECIFIC | 0x04, 0, 0, 0]);
 
     /// The tag context specific tag [5].
-    pub const CTX_5: Self = Tag(Tag::CONTEXT_SPECIFIC | 5);
+    pub const CTX_5: Self = Tag([Tag::CONTEXT_SPECIFIC | 0x05, 0, 0, 0]);
 
     /// The tag context specific tag [6].
-    pub const CTX_6: Self = Tag(Tag::CONTEXT_SPECIFIC | 6);
+    pub const CTX_6: Self = Tag([Tag::CONTEXT_SPECIFIC | 0x06, 0, 0, 0]);
 }
 
 impl Tag {
+    /// Encodes a number into the identifier representation. There are two forms:
+    /// * low tag number (for tag numbers between 0 and 30):
+    ///     One octet. Bits 8 and 7 specify the class , bit 6 indicates whether the encoding
+    ///     is primitive (0), and bits 5-1 give the tag number.
+    /// * High tag number (for tag numbers 31 and greater):
+    ///     Two or more octets. First octet is as in low-tag-number form, except that bits 5-1 all
+    ///     have value 1.
+    ///     Second and following octets give the tag number, base 128, most significant digit first,
+    ///     with as few digits as possible, and with the bit 8 of each octet except the last set
+    ///     to 1.
+    //
+    /// # Panics
+    ///
+    /// Currently, this function panics if the tag number is greater than
+    /// `MAX_VAL_SPAN_3_OCTETS`.
+    #[inline]
+    fn new(number: u32, type_mask: u8) -> Self {
```

Comment:
I think we can make this and all the other constructors const fns now.

Also, could you flip the two arguments, having the `type_mask` (perhaps as `class_mask`) first since the class appears first?

Reply:
> I think we can make this and all the other constructors const fns now.
From what I saw it's not yet stable. Requires `#![feature(const_fn)]`

> Also, could you flip the two arguments, having the type_mask (perhaps as class_mask) first since the class appears first?
Done.

Reply:
A subset of cons fn is stable since 1.31. I think this should be enough here?

Reply:
I tried now -- :(
`error[E0723]: loops and conditional expressions are not stable in const fn`

Reply:
Ah, pitty. But now I wonder if this can be done with some clever bit shifting.

Reply:
I don't think we are far from seeing const fn with conditionals, given that it's already in nightly.
I expect the function could be updated in a release or two...

Reply:
True. I’ll think about it tomorrow for a bit and if I can’t come up with something merge the PR.

Reply:
I concede defeat. We could make the conversion work, but the assertion that the number isn’t greater than 0x1F_FFFF seems impossible.

Reply:
:(. It will sort itself out with const fn soon.